---
title: "Whole-brain comparison of rodent and human brains using spatial transcriptomics"
subtitle: "Results 4, Figure 6"
author: "Antoine Beauchamp"
date: "February 2nd, 2022"
output: html_document
---

#Initialization

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      include = FALSE,
                      cache.lazy = FALSE)
```

```{r libraries}
#Libraries
suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(grid))
suppressPackageStartupMessages(library(gridExtra))
suppressPackageStartupMessages(library(pheatmap))
suppressPackageStartupMessages(library(ggplotify))
suppressPackageStartupMessages(library(ggnewscale))
suppressPackageStartupMessages(library(RMINC))
suppressPackageStartupMessages(library(MRIcrotome))
suppressPackageStartupMessages(library(data.tree))
suppressPackageStartupMessages(library(RColorBrewer))
suppressPackageStartupMessages(library(viridis))
```

```{r functions}
#Functions
source("../../../functions/tree_tools.R")
source("../../../functions/buildSimilarityMatrix.R")
```

```{r import}
#Import data
dfExprMouse <- suppressMessages(read_csv("../../../data/MouseExpressionMatrix_voxel_coronal_maskcoronal_log2_grouped_imputed_labelled.csv"))
dfExprHuman <- suppressMessages(read_csv("../../../data/HumanExpressionMatrix_samples_pipeline_v1_labelled.csv"))

#Extract gene names
genes <- colnames(dfExprMouse)[!str_detect(colnames(dfExprMouse), "Region")]

#Extract mouse voxel labels
dfLabelsMouse <- dfExprMouse %>% 
  select(contains("Region")) %>% 
  mutate(VoxelID = str_c("V", 1:nrow(.)))

#Extract mouse samples labels
dfLabelsHuman <- dfExprHuman %>% 
  select(contains("Region")) %>% 
  mutate(SampleID = str_c("S", 1:nrow(.)))

#Remove expression data frames
rm(dfExprMouse, dfExprHuman)

#Load mouse/human labels (Objects: listLabelsMouseReordered, listLabelsHumanReordered)
load("../../../data/TreeLabelsReordered.RData")

#Import MICe data tree
load("../../../AMBA/data/MouseExpressionTree_DSURQE.RData")
treeMouse <- Clone(treeMouseExpr)
rm(treeMouseExpr)

#Remove white matter and ventricles
pruneAnatTree(treeMouse,
              nodes = c("fiber tracts", "ventricular systems"),
              method = "AtNode")

#Import AHBA data tree
load("../../../AHBA/data/HumanExpressionTree.RData")
treeHuman <- Clone(treeHumanExpr)
rm(treeHumanExpr)

#Fix human colour names (need to start with #)
treeHuman$Do(function(node){
  node$color_hex_triplet <- str_c("#", node$color_hex_triplet)
})

#Remove white matter and ventricles
pruneAnatTree(treeHuman,
              nodes = c("white matter", "sulci & spaces"),
              method = "AtNode")

#Import DSURQE images in AMBA space
dsurqeLabels_200um <- mincGetVolume("../../../AMBA/data/imaging/DSURQE_CCFv3_labels_200um.mnc")
dsurqeMask_200um <- mincGetVolume("../../../AMBA/data/imaging/coronal_200um_coverage_bin0.8.mnc")
dsurqeAnat_200um <- mincGetVolume("../../../AMBA/data/imaging/DSURQE_CCFv3_average_200um.mnc")

#Create a correspondence between voxels in the image space and in the expression matrices

#Indices for the mask and for grey matter ROIs
indMask <- dsurqeMask_200um == 1
indGM <- dsurqeLabels_200um %in% treeMouse$`Basic cell groups and regions`$label
indVoxels <- indMask & indGM
rm(indGM, indMask)

#Create an empty array and assign IDs to non-null voxels
emptyArray <- numeric(length(dsurqeLabels_200um))
names(emptyArray) <- character(length(emptyArray))
voxelNames <- str_c("V", 1:sum(indVoxels))
names(emptyArray)[indVoxels] <- voxelNames
rm(indVoxels)
```

# Figure 6

## Panel A: Similarity of human striatal regions to the mouse striatum

```{r results4-fig6-panelA-import-humansample-sim}
#Specify mouse and human striatal regions based on Balsters et al. 2020
striatumHuman <- c("caudate nucleus",
                   "putamen",
                   "nucleus accumbens")
striatumMouse <- c("Caudoputamen",
                   "Nucleus accumbens")

#ROI labels for human striatum samples
dfLabelsHumanStriatum <- dfLabelsHuman %>% 
  select(SampleID, Region = Region88) %>% 
  filter(Region %in% striatumHuman) 

#List of mouse ROIs
mouseROIs <- dfLabelsMouse %>% 
  select(VoxelID, Region67, Region134) %>% 
  mutate(Region = ifelse(Region67 == "Striatum ventral region", Region134, Region67)) %>% 
  pull(Region) %>% 
  unique() %>% 
  sort()

#Compute the similarity matrix between human striatal samples and mouse ROIs
#for every latent space generated by the MLP
#Run time is about 30 mins
simfile <- "StriatumSimilarityMatrices_HumanSamples.RData"
if (file.exists(simfile)){
  load(simfile)
} else {
  
  #Get the names of latent space voxel and samples matrices
  files <- c("Mouse" = "MouseVoxelTx", "Human" = "HumanVoxelTx") %>% 
    map(function(x){
      str_subset(list.files("../../../data/MLP_outcomes/", full.names = T), x)
    })
  
  #' Title
  #'
  #' @param mouse Name of file containing mouse voxelwise latent space matrix
  #' @param human Name of file containing human samplewise latent space matrix
  #'
  #' @return Similarity matrix for human striatal samples and mouse striatal ROIs
  computeSimilarityMatrix <- function(mouse, human){
    
    #Import mouse voxelwise latent space matrix and compute average hidden unit values
    #for mouse ROIs
    matMouse <- suppressMessages(read_csv(mouse, progress = F)) %>% 
      select(-Region) %>% 
      mutate(VoxelID = str_c("V", 1:nrow(.))) %>% 
      inner_join(dfLabelsMouse %>% 
                   select(VoxelID, Region67, Region134),
                 by = "VoxelID") %>% 
      mutate(Region = ifelse(Region67 == "Striatum ventral region", Region134, Region67)) %>% 
      select(-VoxelID, -Region67, -Region134) %>% 
      group_by(Region) %>% 
      summarise_all(mean) %>% 
      ungroup() %>% 
      column_to_rownames("Region") %>% 
      as.matrix() %>% 
      t()
    
    #Import human samplewise latent space matrix and filter for samples in the striatum
    matHuman <- suppressMessages(read_csv(human, progress = F)) %>% 
      mutate(SampleID = str_c("S", 1:nrow(.))) %>%
      filter(Region %in% striatumHuman) %>% 
      select(-Region) %>% 
      column_to_rownames("SampleID") %>% 
      as.matrix() %>% 
      t()
    
    #Compute the similarity matrix
    matSim <- buildSimilarityMatrix(x1 = matHuman,
                                    x2 = matMouse,
                                    method = "correlation")
    
    return(matSim)
  }
  
  #Map the function over all latent space matrices
  ti <- Sys.time()
  listSimStriatum <- map2(.x = files[["Mouse"]],
                          .y = files[["Human"]],
                          .f = computeSimilarityMatrix)
  names(listSimStriatum) <- str_c("MLP", 1:length(listSimStriatum))
  
  #Save the resulting list to file
  save(listSimStriatum,
       file = simfile)
  
  tf <- Sys.time()
  
  print(tf-ti)
}

makeSimilarityDataFrame <- function(sim){
  out <- sim %>% 
    as_tibble(rownames = "SampleID") %>% 
    inner_join(dfLabelsHuman %>% select(SampleID, Region88),
               by = "SampleID") %>% 
    rename(HumanRegion = Region88) %>% 
    pivot_longer(cols = -c(SampleID, HumanRegion),
                 names_to = "MouseRegion",
                 values_to = "Similarity")
  return(out)
}

#Convert similarity matrices to long form data frames
dfSimStriatum <- map_dfr(.x = listSimStriatum,
                         .f = makeSimilarityDataFrame,
                         .id = "Data")
rm(listSimStriatum)

#Number of MLP samples
nMLPSamples <- length(unique(dfSimStriatum$Data))
```

```{r results4-fig6-panelA-top-mouse-targets}
#Identify mouse targets that are either maximal or in the striatum
dfSimStriatum_MaxSim <- dfSimStriatum %>% 
  group_by(Data, SampleID) %>% 
  mutate(IsMaximal = Similarity == max(Similarity)) %>% 
  ungroup() %>% 
  mutate(IsStriatal = MouseRegion %in% striatumMouse) %>% 
  filter(IsMaximal | IsStriatal)

#Data frame to indicate striatum matches
dfStriatalMatch <- tibble(HumanRegion = striatumHuman,
                          MouseRegion = c("Caudoputamen", "Caudoputamen", "Nucleus accumbens"))

#For every sample in the striatum, compute the average maximal similarity to the expected match
dfSimStriatum_MaxSim_Mean <- dfSimStriatum_MaxSim %>% 
  semi_join(dfStriatalMatch, 
            by = c("HumanRegion", "MouseRegion")) %>% 
  group_by(SampleID) %>% 
  summarise(MeanSimilarity = mean(Similarity)) %>% 
  ungroup()

#Data frame containing bins for proportion values
dfBins <- tibble(BinID = 1:5,
                 BinLabel = c("[0.0, 0.2]",
                              "(0.2, 0.4]",
                              "(0.4, 0.6]",
                              "(0.6, 0.8]",
                              "(0.8, 1.0]")) %>% 
  mutate(BinLabel = factor(BinLabel, levels = BinLabel[order(BinID)]))

#Create a data frame with all combinations of human samples and mouse target regions
dfCombinationGrid <- expand_grid(SampleID = unique(dfSimStriatum_MaxSim$SampleID),
                                 MouseRegion = unique(dfSimStriatum_MaxSim$MouseRegion)) %>% 
  inner_join(dfLabelsHumanStriatum,
             by = "SampleID") %>% 
  rename(HumanRegion = Region) 

#For every sample, compute the proportion of latent spaces where the top hit is the expected match
dfSimStriatum_MaxSim_Proportions <- dfSimStriatum_MaxSim %>% 
  filter(IsMaximal) %>% 
  group_by(SampleID, HumanRegion, MouseRegion) %>% 
  summarise(Prop = n()/nMLPSamples) %>% 
  ungroup() %>% 
  right_join(dfCombinationGrid,
             by = c("SampleID", "HumanRegion", "MouseRegion")) %>% 
  mutate(Prop = ifelse(is.na(Prop), 0, Prop),
         BinID = case_when(Prop <= 0.20 ~ 1,
                           Prop > 0.2 & Prop <= 0.4 ~ 2,
                           Prop > 0.4 & Prop <= 0.6 ~ 3,
                           Prop > 0.6 & Prop <= 0.8 ~ 4,
                           Prop > 0.8 ~ 5)) %>% 
  left_join(dfBins,
            by = "BinID") 

# inner_join(dfSimStriatum_MaxSim_Proportions,
#            dfSimStriatum_MaxSim_Mean,
#            by = "SampleID") %>% 
#   mutate(HumanRegion = as.character(HumanRegion)) %>% 
#   filter(MeanSimilarity > 0.6) %>% 
#   group_by(HumanRegion, MouseRegion) %>% 
#   summarise(MeanProp = mean(Prop)) %>% 
#   ungroup() %>% 
#   filter(MeanProp != 0) %>% 
#   group_by(HumanRegion) %>% 
#   top_n(n = 5, wt = MeanProp) %>% 
#   mutate(MeanProp = round(MeanProp, 4)) %>% 
#   arrange(desc(MeanProp), .by_group = TRUE)
```

```{r results4-fig6-panelA-histogram}
#Specify mouse targets to examine
mouseTargets <- c("Caudoputamen", "Nucleus accumbens", "Fundus of striatum", "Olfactory tubercle")
lvlsHuman <- striatumHuman %>% 
  str_replace("caudate nucleus", "caudate") %>% 
  str_to_sentence()

#For every pair of human striatum regions and mouse targets, get the similarity value for the maximum voxel in every latent space
dfSimStriatum_Distributions <- dfSimStriatum %>% 
  filter(MouseRegion %in% mouseTargets) %>% 
  group_by(Data, HumanRegion, MouseRegion) %>% 
  summarise(MaxPerPair = max(Similarity),
            MeanPerPair = mean(Similarity)) %>% 
  ungroup() %>% 
  mutate(HumanRegion = ifelse(HumanRegion == "caudate nucleus", "caudate", HumanRegion),
         HumanRegion = factor(str_to_sentence(HumanRegion), levels = lvlsHuman),
         MouseRegion = factor(MouseRegion, levels = mouseTargets))

fig6_histogram <- ggplot(dfSimStriatum_Distributions, aes(x = MeanPerPair, y = ..count../nMLPSamples, fill = MouseRegion)) + 
  geom_histogram(col = "grey30",
                 bins = 20,
                 position = "dodge") +
  facet_wrap(~HumanRegion, ncol = length(striatumHuman)) + 
  coord_cartesian(xlim = c(0.6, 1),
                  ylim = c(0,1)) + 
  scale_fill_manual(values = brewer.pal(n = 9, name = "PuBu")[seq(9, 3, by = -2)]) + 
  labs(x = "Average correlation over human samples",
       y = "Proportion of latent spaces",
       fill = "Mouse targets") + 
  theme_bw() + 
  theme(axis.title = element_text(size = 10),
        axis.text = element_text(size = 8),
        legend.position = "bottom",
        plot.margin = margin(b = 0, l = 0, r = 0, t = 0, unit = 'inch'))

#Convert ggplot to grob
fig6_histogram_grob <- ggplotGrob(fig6_histogram) %>% grid.force()

#Extract legend from grob
fig6_histogram_legend_grob <- getGrob(fig6_histogram_grob, "guides.3-3-3-3")

#Remove legend from main plot and make new grob
fig6_histogram <- fig6_histogram +
  theme(legend.position = "none")
fig6_histogram_grob <- ggplotGrob(fig6_histogram) %>% grid.force()

#Panel A title grob
fig6_histogram_title_grob <- textGrob("Similarity of human striatal regions to targets in the mouse striatum",
                                      x = 0,
                                      y = 0.5,
                                      just = c("left", "centre"))
```

## Panels B and C: Voxel-wise brain maps

```{r results4-fig6-panelsBC-import-mousevoxel-sim}
#Human striatal regions
striatumHuman <- c("caudate nucleus",
                   "putamen",
                   "nucleus accumbens")

#Prune the AMBA tree down to 134 bilateral regions
treeMouse_134 <- Clone(treeMouse)
pruneAnatTree(treeMouse_134,
              nodes = unique(dfLabelsMouse$Region134),
              method = "BelowNode")

#Extract striatum ROIs from the tree
striatumMouse <- FindNode(treeMouse_134, "Striatum")$Get("name", filterFun = isLeaf)
names(striatumMouse) <- NULL


#Run time is about 40 minutes
simfile <- "StriatumSimilarityMatrices_MouseVoxels.RData"
if (file.exists(simfile)) {
  load(simfile)
} else {
  
  files <- c("Mouse" = "MouseVoxelTx", "Human" = "HumanVoxelTx") %>% 
    map(function(x){
      str_subset(list.files("../../../data/MLP_outcomes/", full.names = T), x)
    })
  
  computeStriatumSimilarity <- function(mouse, human){
    
    matMouse <- suppressMessages(read_csv(mouse, progress = F)) %>% 
      select(-Region) %>% 
      mutate(VoxelID = str_c("V", 1:nrow(.))) %>% 
      inner_join(dfLabelsMouse %>% 
                   select(VoxelID, Region = Region134),
                 by = "VoxelID") %>% 
      filter(Region %in% striatumMouse) %>% 
      select(-Region) %>% 
      column_to_rownames("VoxelID") %>% 
      as.matrix() %>% 
      t()
    
    matHuman <- suppressMessages(read_csv(human, progress = F)) %>% 
      group_by(Region) %>% 
      summarise_all(.funs = mean) %>% 
      ungroup() %>% 
      column_to_rownames("Region") %>% 
      as.matrix() %>% 
      t()
    
    #Compute similarity matrix
    matSim <- buildSimilarityMatrix(x1 = matHuman,
                                    x2 = matMouse,
                                    method = "correlation")
    
    return(matSim)
  }
  
  nVoxels <- dfLabelsMouse %>% 
    filter(Region134 %in% striatumMouse) %>% 
    nrow()
  
  nROIs <- length(listLabelsHumanReordered$Region88_reordered) 
  
  arraySimStriatum <- array(data = 0,
                            dim = c(nROIs, nVoxels, length(files$Mouse)))
  
  ti <- Sys.time()
  for(i in 1:length(files$Mouse)){
    if(i %% 10 == 0){
      print(i)
    }
    mouse <- files$Mouse[i]
    human <- files$Human[i]
    matSim <- computeStriatumSimilarity(mouse, human)
    arraySimStriatum[ , , i] <- matSim 
  }
  dimnames(arraySimStriatum) <- list(rownames(matSim), colnames(matSim), str_c("MLP", 1:length(files$Mouse)))
  
  save(arraySimStriatum,
       file = simfile)
  
  tf <- Sys.time()
  
  print(tf-ti)
  
}
```

```{r results4-fig6-panelsBC-compute-max-similarity}
#For every latent space, compute the max similarity of every mouse striatal voxel
#Done using a for loop because purrr functions overload memory
listSimStriatum_MaxSim <- vector(mode = "list", length = dim(arraySimStriatum)[3])
for(k in 1:dim(arraySimStriatum)[3]){
  if(k %% 50 == 0){print(k)}
  
  listSimStriatum_MaxSim[[k]] <- arraySimStriatum[,,k] %>%
    as_tibble(rownames = "HumanRegion") %>% 
    pivot_longer(cols = -HumanRegion, names_to = "VoxelID", values_to = "Similarity") %>% 
    group_by(VoxelID) %>% 
    filter(Similarity == max(Similarity)) %>% 
    ungroup() %>% 
    mutate(Data = str_c("MLP_", k))
}

#Reduce list into a data frame
dfSimStriatum_MaxSim <- reduce(listSimStriatum_MaxSim, bind_rows)
rm(listSimStriatum_MaxSim)
```

```{r results4-fig6-panelsBC-top-human-targets}
#Combinations of every mouse voxel and human target
dfCombinations <- expand_grid(VoxelID = unique(dfSimStriatum_MaxSim$VoxelID),
                              HumanRegion = unique(dfSimStriatum_MaxSim$HumanRegion))

#Compute the proportion of latent spaces for which human targets are the maximal match
dfSimStriatum_MaxSim_Proportions <- dfSimStriatum_MaxSim %>% 
  group_by(VoxelID, HumanRegion) %>% 
  summarise(Prop = n()/dim(arraySimStriatum)[3]) %>% 
  ungroup() %>% 
  right_join(dfCombinations, by = c("VoxelID", "HumanRegion")) %>% 
  mutate(Prop = ifelse(is.na(Prop), 0, Prop)) %>% 
  inner_join(dfLabelsMouse %>% select(VoxelID, MouseRegion = Region134),
             by = "VoxelID")

#Identify the human ROIs with the top 5 highest proportions of latent spaces
dfSimStriatum_MaxSim_Proportions_TopHits <- dfSimStriatum_MaxSim_Proportions %>% 
  filter(Prop != 0) %>% 
  group_by(VoxelID) %>% 
  top_n(n = 5, wt = Prop) %>% 
  ungroup()

#Identify the most commonly occurring human regions for voxels in the mouse caudoputamen and nucleus accumbens
# dfSimStriatum_MaxSim_Proportions_TopHits %>% 
#   group_by(MouseRegion, HumanRegion) %>% 
#   summarise(MeanProp = mean(Prop)) %>% 
#   top_n(n = 5, wt = MeanProp) %>% 
#   arrange(MouseRegion, desc(MeanProp))
```

```{r results4-fig6-panelsBC-make-minc-arrays}
#Specify human targets
humanTargets <- c("caudate nucleus", 
                  "putamen",
                  "nucleus accumbens",
                  "substantia innominata",
                  "septal nuclei",
                  "amygdala")

#Compute mean similarity over latent spaces between each voxel and each human region
dfSimStriatumMean <- rowMeans(arraySimStriatum, dims = 2) %>% 
  as_tibble(rownames = "HumanRegion") %>% 
  pivot_longer(cols = -HumanRegion,
               names_to = "VoxelID",
               values_to = "Similarity") %>% 
  inner_join(dfLabelsMouse %>% select(VoxelID, Region134),
             by = "VoxelID") %>% 
  rename(MouseRegion = Region134)

#Filter mean similarity data frame for striatal human targets
dfSimStriatumMean_HumanTargets <- dfSimStriatumMean %>% 
  filter(HumanRegion %in% humanTargets) %>% 
  select(-MouseRegion) %>% 
  pivot_wider(id_cols = VoxelID, names_from = HumanRegion, values_from = Similarity) %>% 
  arrange(VoxelID) %>% 
  column_to_rownames("VoxelID")

#Extract the proportion of max similarity to the selected human targets
dfSimStriatum_MaxSim_Proportions_HumanTargets <- dfSimStriatum_MaxSim_Proportions %>% 
  select(-MouseRegion) %>% 
  filter(HumanRegion %in% humanTargets) %>% 
  pivot_wider(id_cols = "VoxelID", names_from = HumanRegion, values_from = Prop) %>% 
  arrange(VoxelID) %>% 
  column_to_rownames("VoxelID")

#Initialize the MINC arrays for visualization at 200um
listArrayHumanTargets_Mean_200um <- vector(mode = "list", length = ncol(dfSimStriatumMean_HumanTargets))
names(listArrayHumanTargets_Mean_200um) <- colnames(dfSimStriatumMean_HumanTargets)
listArrayHumanTargets_Mean_200um <- map(listArrayHumanTargets_Mean_200um, function(x){return(emptyArray)})

listArrayHumanTargets_Prop_200um <- vector(mode = "list", length = ncol(dfSimStriatum_MaxSim_Proportions_HumanTargets))
names(listArrayHumanTargets_Prop_200um) <- colnames(dfSimStriatum_MaxSim_Proportions_HumanTargets)
listArrayHumanTargets_Prop_200um <- map(listArrayHumanTargets_Prop_200um, function(x){return(emptyArray)})

#For each of the human targets, map the proportions to voxels in a MINC array
for(j in 1:ncol(dfSimStriatumMean_HumanTargets)){
  indVoxelMatch <- match(rownames(dfSimStriatumMean_HumanTargets), names(emptyArray))
  listArrayHumanTargets_Mean_200um[[j]][indVoxelMatch] <- dfSimStriatumMean_HumanTargets[,j]
  listArrayHumanTargets_Prop_200um[[j]][indVoxelMatch] <- dfSimStriatum_MaxSim_Proportions_HumanTargets[,j]
  attributes(listArrayHumanTargets_Mean_200um[[j]]) <- attributes(dsurqeLabels_200um)
  attributes(listArrayHumanTargets_Prop_200um[[j]]) <- attributes(dsurqeLabels_200um)
}

#Paths and commands to resample to 50um
outfiles_mean <- str_c("Figure6_ss_mean_", str_remove(names(listArrayHumanTargets_Mean_200um), " "), "_200um.mnc")
outfiles_prop <- str_c("Figure6_ss_prop_", str_remove(names(listArrayHumanTargets_Prop_200um), " "), "_200um.mnc")
outpaths_mean <- outfiles_mean
outpaths_prop <- outfiles_prop
anatomy50um <- "../../../AMBA/data/imaging/average_template_50um.mnc"
newpaths_mean <- str_replace(outpaths_mean, "200um", "50um")
newpaths_prop <- str_replace(outpaths_prop, "200um", "50um")
resampleCommands_mean <- str_c("mincresample",
                               "-like",
                               anatomy50um,
                               outpaths_mean,
                               newpaths_mean,
                               "-clobber",
                               "-nearest_neighbour",
                               sep = " ")
resampleCommands_prop <- str_c("mincresample",
                               "-like",
                               anatomy50um,
                               outpaths_prop,
                               newpaths_prop,
                               "-clobber",
                               "-nearest_neighbour",
                               sep = " ")

#Resample the images to 50um
dsurqeMask_50um <- mincGetVolume("../../../AMBA/data/imaging/average_template_50um_mask.mnc")
dsurqeAverage_50um <- mincGetVolume("../../../AMBA/data/imaging/DSURQE_CCFv3_average_50um.mnc")
dsurqeAverage_50um[dsurqeMask_50um == 0] <- 0

listArrayHumanTargets_Mean_50um <- vector(mode = "list", length(listArrayHumanTargets_Mean_200um))
listArrayHumanTargets_Prop_50um <- vector(mode = "list", length(listArrayHumanTargets_Prop_200um))
names(listArrayHumanTargets_Mean_50um) <- names(listArrayHumanTargets_Mean_200um)
names(listArrayHumanTargets_Prop_50um) <- names(listArrayHumanTargets_Prop_200um)
for(i in 1:length(listArrayHumanTargets_Mean_200um)){
  
  mincWriteVolume(listArrayHumanTargets_Mean_200um[[i]],
                  output.filename = outpaths_mean[i],
                  like.filename = attributes(listArrayHumanTargets_Mean_200um[[i]])$likeVolume,
                  clobber = TRUE)
  mincWriteVolume(listArrayHumanTargets_Prop_200um[[i]],
                  output.filename = outpaths_prop[i],
                  like.filename = attributes(listArrayHumanTargets_Prop_200um[[i]])$likeVolume,
                  clobber = TRUE)
  
  system(resampleCommands_mean[i])
  system(resampleCommands_prop[i])
  
  listArrayHumanTargets_Mean_50um[[i]] <- mincGetVolume(newpaths_mean[i])
  listArrayHumanTargets_Prop_50um[[i]] <- mincGetVolume(newpaths_prop[i])
  
  attributes(listArrayHumanTargets_Mean_50um[[i]]) <- attributes(dsurqeAverage_50um)
  attributes(listArrayHumanTargets_Prop_50um[[i]]) <- attributes(dsurqeAverage_50um)
}
```

```{r results4-fig6-panelsBC-sliceseries-base}
#Labels for human targets
humanTargets_labels <- c("Caudate",
                         "Putamen",
                         "Nucleus\naccumbens",
                         "Substantia\ninnominata")

#Overlay ranges
avgRange <- c(0.7, 1)
propRange <- c(0.1, 1)

#Palette for average similarity maps
avgPalette <- magma(n = 255)

#Palette for proportion maps
propPalette <- viridis(n = 255)

#Base slice series for average similarity
fig6_ss_avgcorrelation_base <- sliceSeries(nrow = 5, ncol = 1, begin = 144, end = 192) %>% 
  anatomy(mincArray(dsurqeAverage_50um), low = 700, high = 1400) %>% 
  overlay(mincArray(listArrayHumanTargets_Mean_50um[["caudate nucleus"]]), low = avgRange[1], high = avgRange[2], col = avgPalette) %>% 
  sliceSeries() %>% anatomy() %>% 
  overlay(mincArray(listArrayHumanTargets_Mean_50um[["putamen"]]), low = avgRange[1], high = avgRange[2], col = avgPalette) %>% 
  sliceSeries() %>% anatomy() %>% 
  overlay(mincArray(listArrayHumanTargets_Mean_50um[["nucleus accumbens"]]), low = avgRange[1], high = avgRange[2], col = avgPalette) %>% 
  sliceSeries() %>% anatomy() %>% 
  overlay(mincArray(listArrayHumanTargets_Mean_50um[["substantia innominata"]]), low = avgRange[1], high = avgRange[2], col = avgPalette) 

#Base slice series for proportions
fig6_ss_prop_base <- sliceSeries(nrow = 5, ncol = 1, begin = 144, end = 192) %>% 
  anatomy(mincArray(dsurqeAverage_50um), low = 700, high = 1400) %>% 
  overlay(mincArray(listArrayHumanTargets_Prop_50um[["caudate nucleus"]]), low = propRange[1], high = propRange[2], col = propPalette) %>% 
  sliceSeries() %>% anatomy() %>% 
  overlay(mincArray(listArrayHumanTargets_Prop_50um[["putamen"]]), low = propRange[1], high = propRange[2], col = propPalette) %>% 
  sliceSeries() %>% anatomy() %>% 
  overlay(mincArray(listArrayHumanTargets_Prop_50um[["nucleus accumbens"]]), low = propRange[1], high = propRange[2], col = propPalette) %>% 
  sliceSeries() %>% anatomy() %>% 
  overlay(mincArray(listArrayHumanTargets_Prop_50um[["substantia innominata"]]), low = propRange[1], high = propRange[2], col = propPalette)

#Convert slice series to grobs
fig6_ss_avgcorrelation_base_grob <- grobify(fig6_ss_avgcorrelation_base)
fig6_ss_prop_base_grob <- grobify(fig6_ss_prop_base)
```

```{r results4-fig6-panelsBC-sliceseries-legends}
#Create custom legend for mean similarity maps
#Legend scale parameters
fig6_ss_legend_mean_palette <- matrix(avgPalette, nrow = 1, ncol = 255)
fig6_ss_legend_scale_width <- 1
fig6_ss_legend_scale_height <- 0.2
fig6_ss_legend_scale_x <- 0
fig6_ss_legend_scale_y <- 0.95

#Mean similarity legend scale grob
fig6_ss_legend_mean_scale_grob <- rasterGrob(fig6_ss_legend_mean_palette,
                                             x = fig6_ss_legend_scale_x,
                                             y = fig6_ss_legend_scale_y,
                                             width = fig6_ss_legend_scale_width,
                                             height = fig6_ss_legend_scale_height,
                                             just = c("left", "top"))

#Mean similarity legend text
fig6_ss_legend_text_x <- c(0, 1)
fig6_ss_legend_text_y <- fig6_ss_legend_scale_y - fig6_ss_legend_scale_height - 0.05
fig6_ss_legend_mean_text_grob <- textGrob(as.character(avgRange),
                                          x = fig6_ss_legend_text_x,
                                          y = fig6_ss_legend_text_y,
                                          just = c("center", "top"),
                                          gp = gpar(fontsize = 8))

#Mean similarity legend title
fig6_ss_legend_title_x <- (fig6_ss_legend_scale_x + fig6_ss_legend_scale_width)/2
fig6_ss_legend_title_y <- fig6_ss_legend_text_y - 0.2
fig6_ss_legend_mean_title_grob <- textGrob("Average correlation",
                                           x = fig6_ss_legend_title_x,
                                           y = fig6_ss_legend_title_y,
                                           just = c("center", "top"),
                                           gp = gpar(fontsize = 10))

#Combine mean similarity legend grobs into one
fig6_ss_legend_mean_grob <- gTree(children = gList(fig6_ss_legend_mean_scale_grob,
                                                   fig6_ss_legend_mean_text_grob,
                                                   fig6_ss_legend_mean_title_grob))

#Create custom legend for proportion maps
#Proportion legend scale
fig6_ss_legend_prop_palette <- matrix(propPalette, nrow = 1, ncol = 255)
fig6_ss_legend_prop_scale_grob <- rasterGrob(fig6_ss_legend_prop_palette,
                                             x = fig6_ss_legend_scale_x,
                                             y = fig6_ss_legend_scale_y,
                                             width = fig6_ss_legend_scale_width,
                                             height = fig6_ss_legend_scale_height,
                                             just = c("left", "top"))

#Proportion legend text
fig6_ss_legend_prop_text_grob <- textGrob(as.character(propRange),
                                          x = fig6_ss_legend_text_x,
                                          y = fig6_ss_legend_text_y,
                                          just = c("center", "top"),
                                          gp = gpar(fontsize = 8))

#Proportion legend title
fig6_ss_legend_prop_title_grob <- textGrob("Proportion of gene expression latent spaces",
                                           x = fig6_ss_legend_title_x,
                                           y = fig6_ss_legend_title_y,
                                           just = c("center", "top"),
                                           gp = gpar(fontsize = 10))

#Combine proportion legend grobs into one
fig6_ss_legend_prop_grob <- gTree(children = gList(fig6_ss_legend_prop_scale_grob,
                                                   fig6_ss_legend_prop_text_grob,
                                                   fig6_ss_legend_prop_title_grob))
```

```{r results4-fig6-panelsBC-sliceseries-labels}
#Labels for human targets
humanTargets_labels <- c("Caudate",
                         "Putamen",
                         "Nucleus\naccumbens",
                         "Substantia\ninnominata")

#Create text grob for human target labels for mean similarity maps
fig6_ss_labels_mean_grob <- textGrob(humanTargets_labels,
                                     x = 0.125 + 0.25*(0:3),
                                     y = 0.2,
                                     just = c("center", "bottom"),
                                     gp = gpar(lineheight = 1, fontsize = 8))

#Create text grob for human target labels for proportion maps
fig6_ss_labels_prop_grob <- fig6_ss_labels_mean_grob
```

```{r results4-fig6-panelsBC-sliceseries-titles}
#Panel B title grob
fig6_ss_title_mean_grob <- textGrob("Average similarity to human targets",
                                    x = 0,
                                    y = 0.5,
                                    just = c("left", "centre"))

#Panel C title grob
fig6_ss_title_prop_grob <- textGrob("Voxels maximally similar to human targets",
                                    x = 0,
                                    y = 0.5,
                                    just = c("left", "centre"))
```

## Figure 6 grob

```{r results4-fig6-grob, fig.width = 8, fig.height = 6.8}
#Figure 6 plots grob
fig6_plots_grob <- arrangeGrob(textGrob("A.", gp = gpar(fontsize = 14, fontface = "bold")), #1
                               fig6_histogram_title_grob, #2
                               fig6_histogram_grob, #3
                               fig6_histogram_legend_grob, #4
                               textGrob("B.", gp = gpar(fontsize = 14, fontface = "bold")), #5
                               fig6_ss_title_mean_grob, #6
                               textGrob("C.", gp = gpar(fontsize = 14, fontface = "bold")), #7
                               fig6_ss_title_prop_grob, #6
                               fig6_ss_labels_mean_grob, #6
                               fig6_ss_labels_prop_grob, #7
                               fig6_ss_avgcorrelation_base_grob, #8
                               fig6_ss_prop_base_grob, #9
                               fig6_ss_legend_mean_grob, #10
                               fig6_ss_legend_prop_grob, #11
                               layout_matrix = rbind(c(NA, NA, NA, NA, NA, NA),
                                                     c(NA,  1,  2,  2,  2, NA),
                                                     c(NA,  3,  3,  3,  3, NA),
                                                     c(NA,  4,  4,  4,  4, NA),
                                                     c(NA, NA, NA, NA, NA, NA),
                                                     c(NA,  5,  6,  7,  8, NA),
                                                     c(NA, NA,  9, NA, 10, NA),
                                                     c(NA, NA, 11, NA, 12, NA),
                                                     c(NA, NA, 13, NA, 14, NA),
                                                     c(NA, NA, NA, NA, NA, NA)),
                               widths = unit(c(0.29, 0.46, 3.25, 0.5, 3.25, 0.25), 'inch'),
                               heights = unit(c(0.1, 0.4, 2, 0.4, 0.2, 0.2, 0.4, 2.5, 0.5, 0.1), 'inch'))

# #Grob for figure border
fig6_border_grob <- rectGrob(gp = gpar(fill = NA))

#Figure 6 grob
fig6_grob <- gTree(children = gList(fig6_plots_grob,
                                    fig6_border_grob))

fig6_width <- 8
fig6_height <- 6.8

grid.newpage()
grid.draw(fig6_grob)
```

```{r results4-fig6-write}
fileout <- "Figure6.RData"
save(fig6_grob,
     fig6_width,
     fig6_height,
     file = fileout)
```

# In-text quantification

## Panel A

```{r results4-fig6-intext-1, eval = F}
#Mouse and human ROIs
mouseTargets <- c("Caudoputamen", "Nucleus accumbens", "Fundus of striatum", "Olfactory tubercle")
humanSeeds <- c("Caudate nucleus", "Putamen", "Nucleus accumbens")

#For every pair of human striatum regions and mouse targets, get the similarity value for the maximum voxel in every latent space
dfSimStriatum_Distributions <- dfSimStriatum %>% 
  filter(MouseRegion %in% mouseTargets) %>% 
  group_by(Data, HumanRegion, MouseRegion) %>% 
  summarise(MaxPerPair = max(Similarity),
            MeanPerPair = mean(Similarity)) %>% 
  ungroup() %>% 
  mutate(HumanRegion = factor(str_to_sentence(HumanRegion), levels = str_to_sentence(striatumHuman)),
         MouseRegion = factor(MouseRegion, levels = mouseTargets))

#Summary statistics for mouse-human pairs
dfSimStriatum_Distributions %>% 
  filter(HumanRegion %in% humanSeeds,
         MouseRegion %in% mouseTargets) %>% 
  group_by(MouseRegion, HumanRegion) %>% 
  summarise(DistMean = mean(MeanPerPair),
            DistMedian = median(MeanPerPair),
            DistSd = sd(MeanPerPair)) %>% 
  select(HumanRegion, MouseRegion, starts_with("Dist")) %>% 
  arrange(HumanRegion)
```

```{r results4-fig6-intext-2, eval = F}
#Modal value for caudate-caudoputamen distribution
ggplot_build(fig6_histogram)$data[[1]] %>% 
  filter(group == 1, PANEL == 1) %>% 
  select(y, count, x) %>% 
  arrange(desc(y)) %>% 
  head()
```

```{r results4-fig6-intext-3, eval = F}
#Modal value for putamen-caudoputamen distribution
ggplot_build(fig6_histogram)$data[[1]] %>% 
  filter(group == 1, PANEL == 2) %>% 
  select(y, count, x) %>% 
  arrange(desc(y)) %>% 
  head()
```

```{r results4-fig6-intext-4, eval = F}
dfSimStriatum_Distributions %>% 
  filter(HumanRegion %in% humanSeeds,
         MouseRegion %in% mouseTargets) %>% 
  group_by(MouseRegion, HumanRegion) %>% 
  summarise(thresh0.95 = sum(MeanPerPair >= 0.95)/nMLPSamples,
            thresh0.90 = sum(MeanPerPair >= 0.90)/nMLPSamples,
            thresh0.80 = sum(MeanPerPair >= 0.80)/nMLPSamples) %>% 
  select(HumanRegion, MouseRegion, starts_with("thresh")) %>% 
  arrange(HumanRegion)
```

```{r results4-fig6-intext-5, eval = F}
#Variance calculated over all mouse targets
dfSimStriatum_Distributions %>% 
  filter(HumanRegion %in% humanSeeds,
         MouseRegion %in% mouseTargets) %>% 
  group_by(HumanRegion) %>% 
  summarise(sd(MeanPerPair))
```

## Panel B

```{r results4-fig6-intext-6, eval = F}
#Specify human targets
humanTargets <- c("caudate nucleus", 
                  "putamen",
                  "nucleus accumbens",
                  "septal nuclei",
                  "amygdala")

#Average correlations and standard deviations of mouse caudoputamen voxels to human regions
dfSimStriatumMean %>% 
  filter(MouseRegion == "Caudoputamen") %>% 
  group_by(HumanRegion) %>% 
  summarise(SimMean = mean(Similarity),
            SimSd = sd(Similarity)) %>% 
  arrange(desc(SimMean))
```

```{r results4-fig6-intext-7, eval = F}
#Percentage of caudputamen voxels with mean similarity of at least 0.9 to different human targets
dfSimStriatumMean %>% 
  filter(HumanRegion %in% humanTargets) %>%
  filter(MouseRegion == "Caudoputamen") %>% 
  group_by(HumanRegion) %>% 
  summarise(nFilter = sum(Similarity >= 0.9),
            nCP = n(),
            nFilter/nCP) %>% 
  arrange(desc(nFilter))
```

```{r results4-fig6-intext-8, eval = F}
#Average correlations and standard deviations of mouse accumbens voxels to human regions
dfSimStriatumMean %>% 
  filter(MouseRegion == "Nucleus accumbens") %>% 
  group_by(HumanRegion) %>% 
  summarise(SimMean = mean(Similarity),
            SimSd = sd(Similarity)) %>% 
  arrange(desc(SimMean))
```

```{r results4-fig6-intext-9, eval = F}
#Percentage of accumbens voxels with mean similarity of at least 0.7 to different human targets
dfSimStriatumMean %>% 
  filter(MouseRegion == "Nucleus accumbens") %>% 
  group_by(HumanRegion) %>% 
  summarise(nFilter = sum(Similarity >= 0.7),
            nCP = n(),
            nFilter/nCP) %>% 
  arrange(desc(nFilter))
```

## Panel C

```{r results4-fig6-intext-10, eval = F}
#Extract the proportions to the selected human targets
dfSimStriatum_MaxSim_Proportions_HumanTargets <- dfSimStriatum_MaxSim_Proportions %>% 
  select(-MouseRegion) %>% 
  filter(HumanRegion %in% humanTargets) %>% 
  pivot_wider(id_cols = "VoxelID", names_from = HumanRegion, values_from = Prop) %>% 
  column_to_rownames("VoxelID")

#Percentage of caudoputamen voxels being maximally similar to caudate or putamen in at least 95% of latent spaces and all latent spaces
dfSimStriatum_MaxSim_Proportions %>% 
  filter(HumanRegion %in% humanTargets) %>% 
  filter(MouseRegion == "Caudoputamen") %>% 
  filter(HumanRegion %in% c("caudate nucleus", "putamen")) %>% 
  group_by(VoxelID) %>% 
  summarise(PropSum = sum(Prop)) %>% 
  ungroup %>% 
  summarise(sum(PropSum >= 0.95)/n(),
            sum(PropSum == 1)/n())
```

```{r results4-fig6-intext-11, eval = F}
#Difference in average similarity to caudate and putamen for voxels in the caudoputamen
dfSimStriatumMean %>% 
  filter(HumanRegion %in% c("caudate nucleus", "putamen")) %>% 
  filter(MouseRegion == "Caudoputamen") %>% 
  select(VoxelID, HumanRegion, Similarity) %>% 
  pivot_wider(id_cols = VoxelID, names_from = HumanRegion, values_from = Similarity) %>% 
  mutate(diff = `caudate nucleus` - `putamen`) %>% 
  pull(diff) %>% 
  mean()
```

```{r results4-fig6-intext-12, eval = F}
#Number of voxels per mouse region
dfNumVoxels <- dfSimStriatum_MaxSim_Proportions %>% 
  select(VoxelID, MouseRegion) %>% 
  distinct() %>% 
  group_by(MouseRegion) %>% 
  summarise(nVoxels = n()) %>% 
  ungroup()

#Proportion of voxels in mouse regions having nucleus accumbens as top match in at least 80% of latent spaces
dfSimStriatum_MaxSim_Proportions %>% 
  filter(HumanRegion %in% humanTargets) %>% 
  filter(HumanRegion == "nucleus accumbens") %>% 
  arrange(desc(Prop)) %>% 
  filter(Prop >= 0.8) %>% 
  group_by(MouseRegion) %>% 
  count() %>% 
  ungroup() %>% 
  inner_join(dfNumVoxels, by = "MouseRegion") %>% 
  mutate(Prop = n/nVoxels)
```

```{r results4-fig6-intext-13, eval = F}
#Voxels in the mouse accumbens with prop max similarity to the human accumbens in less than 80% of latent spaces
voxelNAcc <- dfSimStriatum_MaxSim_Proportions %>% 
  filter(HumanRegion %in% humanTargets) %>% 
  filter(HumanRegion == "nucleus accumbens") %>% 
  filter(MouseRegion == "Nucleus accumbens") %>% 
  arrange(desc(Prop)) %>% 
  filter(Prop < 0.8) %>% 
  pull(VoxelID)


dfSimStriatum_MaxSim_Proportions %>% 
  filter(VoxelID %in% voxelNAcc) %>% 
  filter(Prop >= 0.3) %>% 
  group_by(VoxelID) %>% 
  top_n(n = 1, wt = Prop) %>% 
  group_by(HumanRegion) %>% 
  count()
```

```{r results4-fig6-intext-14, eval = F}
#Voxels in the mouse olfactory tubercle with prop max similarity to the human accumbens in less than 80% of latent spaces
voxelOlf <- dfSimStriatum_MaxSim_Proportions %>% 
  filter(HumanRegion %in% humanTargets) %>% 
  filter(HumanRegion == "nucleus accumbens") %>% 
  filter(MouseRegion == "Olfactory tubercle") %>% 
  arrange(desc(Prop)) %>% 
  filter(Prop < 0.8) %>% 
  pull(VoxelID)

dfSimStriatum_MaxSim_Proportions %>% 
  filter(VoxelID %in% voxelOlf) %>% 
  filter(Prop >= 0.3) %>% 
  group_by(VoxelID) %>% 
  top_n(n = 1, wt = Prop) %>% 
  group_by(HumanRegion) %>% 
  count()
```
