{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "923c3057",
   "metadata": {},
   "source": [
    "Allen Institute resources for API access:\n",
    "- http://help.brain-map.org/display/api/Allen+Brain+Atlas+API\n",
    "- http://help.brain-map.org/display/api/Downloading+3-D+Expression+Grid+Data\n",
    "- http://help.brain-map.org/display/api/Example+Queries+for+Experiment+Metadata\n",
    "- http://help.brain-map.org/pages/viewpage.action?pageId=5308449\n",
    "\n",
    "XML parsing:\n",
    "- https://www.datacamp.com/community/tutorials/python-xml-elementtree\n",
    "- https://www.kite.com/python/answers/how-to-download-a-csv-file-from-a-url-in-python\n",
    "\n",
    "Python parallel processing:\n",
    "- https://www.machinelearningplus.com/python/parallel-processing-python\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7dd8a761",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import argparse\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import requests\n",
    "from pyminc.volumes.factory import *\n",
    "from zipfile import ZipFile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ebc542ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_metadata(dataset = 'coronal', outdir='./'):\n",
    "\n",
    "    \"\"\" \"\"\"\n",
    "\n",
    "    metadata_file = outdir+'AMBA_metadata_{}.csv'.format(dataset)\n",
    "\n",
    "    if os.path.isfile(metadata_file):\n",
    "        metadata = pd.read_csv(metadata_file, index_col = None)\n",
    "    else:\n",
    "        abi_query_metadata = \"http://api.brain-map.org/api/v2/data/SectionDataSet/query.csv?\"+\\\n",
    "\"criteria=[failed$eqfalse],plane_of_section[name$eq{}],products[abbreviation$eqMouse],treatments[name$eqISH],genes&\".format(dataset)+\\\n",
    "\"tabular=data_sets.id+as+experiment_id,data_sets.section_thickness,data_sets.specimen_id,\"+\\\n",
    "\"plane_of_sections.name+as+plane,\"+\\\n",
    "\"genes.acronym+as+gene,genes.name+as+gene_name,genes.chromosome_id,genes.entrez_id,genes.genomic_reference_update_id,genes.homologene_id,genes.organism_id&\"+\\\n",
    "\"start_row=0&num_rows=all\"\n",
    "\n",
    "        metadata = pd.read_csv(abi_query_metadata)\n",
    "        metadata.to_csv(metadata_file, index=False)\n",
    "\n",
    "    return metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "8a1fc5b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_expression(experiment_id, outdir = './tmp/'):\n",
    "\n",
    "    \"\"\" \"\"\"\n",
    "\n",
    "    if not os.path.exists(outdir):\n",
    "        os.mkdir(outdir)\n",
    "\n",
    "    abi_query_expr = 'http://api.brain-map.org/grid_data/download/{}'.format(experiment_id)\n",
    "\n",
    "    amba_request = requests.get(abi_query_expr)\n",
    "\n",
    "    tmpfile = outdir+str(experiment_id)+'.zip'\n",
    "    with open(tmpfile, 'wb') as file:\n",
    "        file.write(amba_request.content)\n",
    "\n",
    "    with ZipFile(tmpfile, 'r') as file:\n",
    "        try:\n",
    "            file.extract('energy.raw', path = outdir)\n",
    "            os.rename(outdir+'energy.raw', outdir+str(experiment_id)+'.raw')\n",
    "            success = 1\n",
    "        except KeyError as err:\n",
    "            print('Error for experiment {}: {}'.format(experiment_id, err))\n",
    "            success = 0\n",
    "            \n",
    "    os.remove(tmpfile)\n",
    " \n",
    "    return success"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "71db075c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def transform_space(infile, outfile, voxel_orientation = 'RAS', world_space = 'MICe', expansion_factor = 1.0, volume_type = None, data_type = None, labels = False):\n",
    "\n",
    "    def reorient_to_standard(dat):\n",
    "        dat = np.rot90(dat, k=1, axes=(0, 2))\n",
    "        dat = np.rot90(dat, k=1, axes=(0, 1))\n",
    "\n",
    "        shape = dat.shape\n",
    "        dat = np.ravel(dat)\n",
    "        dat = np.reshape(dat, shape)\n",
    "\n",
    "        return(dat)\n",
    "\n",
    "    def do_nothing(dat):\n",
    "        return(dat)\n",
    "\n",
    "    # %% Coordinate definitions\n",
    "\n",
    "    # Centers are listed as x,y,z; reverse these when writing out\n",
    "    # Centers listed in um in CCFv3 coordinates\n",
    "    centers_RAS = {\"MICe\"   :   [5700, 7900, 2700],\n",
    "                   \"CCFv3\"  :   [0, 13200, 8000]}\n",
    "    centers_PIR = {\"MICe\"   :   [5300, 5300, 5700],\n",
    "                   \"CCFv3\"  :   [0, 0, 0]}\n",
    "\n",
    "    # Direction cosines\n",
    "    direction_cosines_RAS = {\"MICe\"     :   [[1, 0, 0], [0, 1, 0], [0, 0, 1]],\n",
    "                             \"CCFv3\"    :   [[0, 0, 1], [-1, 0, 0], [0, -1, 0]]}\n",
    "\n",
    "    direction_cosines_PIR = {\"MICe\"     :   [[0, -1, 0], [0, 0, -1], [1, 0, 0]],\n",
    "                             \"CCFv3\"    :   [[1, 0, 0], [0, 1, 0], [0, 0, 1]]}\n",
    "\n",
    "    # Map arguments to functions/dicts/values\n",
    "    map_voxel_orientations = {\"RAS\" :   reorient_to_standard,\n",
    "                              \"PIR\":    do_nothing}\n",
    "\n",
    "    map_centers = {\"RAS\"    :   centers_RAS,\n",
    "                   \"PIR\"    :   centers_PIR}\n",
    "\n",
    "    map_dir_cosines = {\"RAS\"    :   direction_cosines_RAS,\n",
    "                       \"PIR\"    :   direction_cosines_PIR}\n",
    "\n",
    "    size_10 = 1320*800*1140\n",
    "    size_25 = 528*320*456\n",
    "    size_50 = 264*160*228\n",
    "    size_100 = 132*80*114\n",
    "    size_200 = 58*41*67\n",
    "\n",
    "    map_resolutions = {size_10: 10,\n",
    "                       size_25: 25,\n",
    "                       size_50: 50,\n",
    "                       size_100: 100,\n",
    "                       size_200: 200}\n",
    "\n",
    "    vol = volumeFromFile(infile)\n",
    "\n",
    "    res = map_resolutions[vol.data.size]\n",
    "\n",
    "    # Voxel orientation\n",
    "    if voxel_orientation in map_voxel_orientations:\n",
    "        new_data = map_voxel_orientations[voxel_orientation](vol.data)\n",
    "    else:\n",
    "        print(\"Invalid voxel orientation\")\n",
    "        sys.exit(1)\n",
    "\n",
    "    # World coordinate system\n",
    "    centers = [expansion_factor*c/(1000) \n",
    "               for c in map_centers[voxel_orientation][world_space]]\n",
    "    steps = [expansion_factor*res/1000] * 3\n",
    "    xdc = map_dir_cosines[voxel_orientation][world_space][0]\n",
    "    ydc = map_dir_cosines[voxel_orientation][world_space][1]\n",
    "    zdc = map_dir_cosines[voxel_orientation][world_space][2]\n",
    "\n",
    "    # Types\n",
    "    vtype = vol.volumeType if volume_type is None else volume_type\n",
    "    dtype = vol.dtype if data_type is None else data_type\n",
    "    labels = vol.labels if labels is None else labels\n",
    "\n",
    "    outvol = volumeFromDescription(outputFilename=outfile,\n",
    "                                   dimnames=[\"zspace\", \"yspace\", \"xspace\"],\n",
    "                                   sizes=new_data.shape,\n",
    "                                   starts=[-c for c in reversed(centers)],\n",
    "                                   steps=[s for s in reversed(steps)],\n",
    "                                   x_dir_cosines=xdc,\n",
    "                                   y_dir_cosines=ydc,\n",
    "                                   z_dir_cosines=zdc,\n",
    "                                   volumeType=vtype,\n",
    "                                   dtype=dtype,\n",
    "                                   labels=labels)\n",
    "\n",
    "    outvol.data = new_data\n",
    "    outvol.writeFile()\n",
    "    outvol.closeVolume()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "5e51515d",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = 'sagittal'\n",
    "datadir = 'data/expression/'\n",
    "\n",
    "dfMetadata = fetch_metadata(dataset = dataset,\n",
    "                           outdir = datadir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "f3eac853",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error for experiment 375: \"There is no item named 'energy.raw' in the archive\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cat: data/expression/sagittal/375.raw: No such file or directory\n",
      "rawtominc: Premature end of file.\n"
     ]
    },
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'data/expression/sagittal/375.raw'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Input \u001b[0;32mIn [22]\u001b[0m, in \u001b[0;36m<cell line: 2>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     14\u001b[0m cmd \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcat \u001b[39m\u001b[38;5;132;01m{}\u001b[39;00m\u001b[38;5;124m | rawtominc \u001b[39m\u001b[38;5;132;01m{}\u001b[39;00m\u001b[38;5;124m -signed -float -ounsigned -oshort -xstep 0.2 -ystep 0.2 -zstep 0.2 -clobber 58 41 67\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m.\u001b[39mformat(infile, outfile)\n\u001b[1;32m     16\u001b[0m os\u001b[38;5;241m.\u001b[39msystem(cmd)\n\u001b[0;32m---> 17\u001b[0m os\u001b[38;5;241m.\u001b[39mremove(infile)\n\u001b[1;32m     19\u001b[0m infile \u001b[38;5;241m=\u001b[39m outfile\n\u001b[1;32m     20\u001b[0m outfile \u001b[38;5;241m=\u001b[39m outdir\u001b[38;5;241m+\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{}\u001b[39;00m\u001b[38;5;124m_tmp2.mnc\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m.\u001b[39mformat(experiment_id)\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'data/expression/sagittal/375.raw'"
     ]
    }
   ],
   "source": [
    "dfTemp = dfMetadata.loc[:5]\n",
    "for i, row in dfTemp.iterrows():\n",
    "\n",
    "    experiment_id = row['experiment_id']\n",
    "\n",
    "    outdir = datadir+'{}/'.format(dataset)\n",
    "    success = fetch_expression(experiment_id = experiment_id, \n",
    "                               outdir = outdir)\n",
    "\n",
    "\n",
    "    infile = outdir+'{}.raw'.format(experiment_id)\n",
    "    outfile = outdir+'{}_tmp.mnc'.format(experiment_id)\n",
    "\n",
    "    cmd = 'cat {} | rawtominc {} -signed -float -ounsigned -oshort -xstep 0.2 -ystep 0.2 -zstep 0.2 -clobber 58 41 67'.format(infile, outfile)\n",
    "\n",
    "    os.system(cmd)\n",
    "    os.remove(infile)\n",
    "\n",
    "    infile = outfile\n",
    "    outfile = outdir+'{}_tmp2.mnc'.format(experiment_id)\n",
    "    transform_space(infile = infile,\n",
    "                    outfile = outfile,\n",
    "                    voxel_orientation = 'RAS',\n",
    "                    world_space = 'MICe',\n",
    "                    expansion_factor = 1.0)\n",
    "    os.remove(infile)\n",
    "\n",
    "\n",
    "    gene_id = row['gene']\n",
    "\n",
    "    infile = outfile\n",
    "    outfile = outdir+'{}_{}.mnc'.format(gene_id, experiment_id)\n",
    "    os.rename(infile, outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "66606ae6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error for experiment 375: \"There is no item named 'energy.raw' in the archive\"\n"
     ]
    }
   ],
   "source": [
    "experiment_id = dfMetadata.loc[0,'experiment_id']\n",
    "\n",
    "outdir = datadir+'{}/'.format(dataset)\n",
    "success = fetch_expression(experiment_id = experiment_id, \n",
    "                 outdir = outdir)\n",
    "\n",
    "if success == 1:\n",
    "\n",
    "    infile = outdir+'{}.raw'.format(experiment_id)\n",
    "    outfile = outdir+'{}_tmp.mnc'.format(experiment_id)\n",
    "\n",
    "    cmd = 'cat {} | rawtominc {} -signed -float -ounsigned -oshort -xstep 0.2 -ystep 0.2 -zstep 0.2 -clobber 58 41 67'.format(infile, outfile)\n",
    "\n",
    "    os.system(cmd)\n",
    "    os.remove(infile)\n",
    "\n",
    "    infile = outfile\n",
    "    outfile = outdir+'{}_tmp2.mnc'.format(experiment_id)\n",
    "    transform_space(infile = infile,\n",
    "                    outfile = outfile,\n",
    "                    voxel_orientation = 'RAS',\n",
    "                    world_space = 'MICe',\n",
    "                    expansion_factor = 1.0)\n",
    "    os.remove(infile)\n",
    "\n",
    "\n",
    "    gene_id = dfMetadata.loc[0,'gene']\n",
    "\n",
    "    infile = outfile\n",
    "    outfile = outdir+'{}_{}.mnc'.format(gene_id, experiment_id)\n",
    "    os.rename(infile, outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "9fc02d46",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfMetadata['success'] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "72efcaf3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>experiment_id</th>\n",
       "      <th>section_thickness</th>\n",
       "      <th>specimen_id</th>\n",
       "      <th>plane</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chromosome_id</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>genomic_reference_update_id</th>\n",
       "      <th>homologene_id</th>\n",
       "      <th>organism_id</th>\n",
       "      <th>success</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>375</td>\n",
       "      <td>25</td>\n",
       "      <td>710147</td>\n",
       "      <td>sagittal</td>\n",
       "      <td>Ebf1</td>\n",
       "      <td>early B cell factor 1</td>\n",
       "      <td>36.0</td>\n",
       "      <td>13591.0</td>\n",
       "      <td>491928275.0</td>\n",
       "      <td>7297.0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>386</td>\n",
       "      <td>25</td>\n",
       "      <td>710148</td>\n",
       "      <td>sagittal</td>\n",
       "      <td>Efnb1</td>\n",
       "      <td>ephrin B1</td>\n",
       "      <td>57.0</td>\n",
       "      <td>13641.0</td>\n",
       "      <td>491928275.0</td>\n",
       "      <td>3263.0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>81600550</td>\n",
       "      <td>25</td>\n",
       "      <td>6510</td>\n",
       "      <td>sagittal</td>\n",
       "      <td>Tmem30a</td>\n",
       "      <td>transmembrane protein 30A</td>\n",
       "      <td>54.0</td>\n",
       "      <td>69981.0</td>\n",
       "      <td>491928275.0</td>\n",
       "      <td>110703.0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>68269598</td>\n",
       "      <td>25</td>\n",
       "      <td>67501035</td>\n",
       "      <td>sagittal</td>\n",
       "      <td>Csrp1</td>\n",
       "      <td>cysteine and glycine-rich protein 1</td>\n",
       "      <td>34.0</td>\n",
       "      <td>13007.0</td>\n",
       "      <td>491928275.0</td>\n",
       "      <td>37874.0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>68844361</td>\n",
       "      <td>25</td>\n",
       "      <td>68151384</td>\n",
       "      <td>sagittal</td>\n",
       "      <td>Slc4a1ap</td>\n",
       "      <td>solute carrier family 4 (anion exchanger), mem...</td>\n",
       "      <td>50.0</td>\n",
       "      <td>20534.0</td>\n",
       "      <td>491928275.0</td>\n",
       "      <td>7543.0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   experiment_id  section_thickness  specimen_id     plane      gene  \\\n",
       "0            375                 25       710147  sagittal      Ebf1   \n",
       "1            386                 25       710148  sagittal     Efnb1   \n",
       "2       81600550                 25         6510  sagittal   Tmem30a   \n",
       "3       68269598                 25     67501035  sagittal     Csrp1   \n",
       "4       68844361                 25     68151384  sagittal  Slc4a1ap   \n",
       "\n",
       "                                           gene_name  chromosome_id  \\\n",
       "0                              early B cell factor 1           36.0   \n",
       "1                                          ephrin B1           57.0   \n",
       "2                          transmembrane protein 30A           54.0   \n",
       "3                cysteine and glycine-rich protein 1           34.0   \n",
       "4  solute carrier family 4 (anion exchanger), mem...           50.0   \n",
       "\n",
       "   entrez_id  genomic_reference_update_id  homologene_id  organism_id  success  \n",
       "0    13591.0                  491928275.0         7297.0            2        0  \n",
       "1    13641.0                  491928275.0         3263.0            2        0  \n",
       "2    69981.0                  491928275.0       110703.0            2        0  \n",
       "3    13007.0                  491928275.0        37874.0            2        0  \n",
       "4    20534.0                  491928275.0         7543.0            2        0  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfMetadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "af8af173",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "success"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "5f09085f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bool(success)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d549d5d5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
