---
title: "Transcriptomic similarity in the mouse and human brain"
subtitle: "Results 2, Figure 3"
author: "Antoine Beauchamp"
date: 'January 25th, 2022'
output: html_document
---

# Initialization

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      include = FALSE,
                      cache.lazy = FALSE)
```

```{r libraries}
#Libraries
suppressPackageStartupMessages(library(tidyverse))
library(grid)
library(gridExtra)
library(pheatmap)
library(ggplotify)
library(ggnewscale)
library(RMINC)
library(MRIcrotome)
library(data.tree)
library(RColorBrewer)
library(viridis)
```

```{r functions}
#Functions
source("/projects/abeauchamp/Functions/TreeTools.R")
source("/projects/abeauchamp/Functions/mincBuildArray.R")
source("/projects/abeauchamp/Projects/MouseHumanMapping/Functions/ProcessingTools.R")
source("/projects/abeauchamp/Projects/MouseHumanMapping/Functions/buildSimilarityMatrix.R")
```

```{r import}
#Set path
pathHome <- "/projects/abeauchamp/Projects/MouseHumanMapping/Paper_Descriptive/"

#Import data
dfExprMouse <- suppressMessages(read_csv(str_c(pathHome, "Data/", "MouseExpressionMatrix_Voxel_coronal_maskcoronal_imputed_labelled.csv")))
dfExprHuman <- suppressMessages(read_csv(str_c(pathHome, "Data/", "HumanExpressionMatrix_Samples_labelled.csv")))

#Load mouse/human labels (Objects: listLabelsMouseReordered, listLabelsHumanReordered)
load(str_c(pathHome, "Data/", "TreeLabelsReordered.RData"))

#Import MICe data tree
load(str_c("/projects/abeauchamp/Projects/MouseHumanMapping/AllenMouseBrainAtlas/Data/MouseExpressionTree_DSURQE.RData"))
treeMouse <- Clone(treeMouseExpr)
rm(treeMouseExpr)

#Remove white matter and ventricles
pruneAnatTree(treeMouse, 
              nodes = c("fiber tracts", "ventricular systems"),
              method = "AtNode")

#Import AHBA data tree
load("/projects/abeauchamp/Projects/MouseHumanMapping/AllenHumanBrainAtlas/Data/HumanExpressionTree_Samples.RData")
treeHuman <- Clone(treeHumanExpr)
rm(treeHumanExpr)

#Fix human colour names (need to start with #)
treeHuman$Do(function(node){
  node$color_hex_triplet <- str_c("#", node$color_hex_triplet)
})

#Remove white matter and ventricles
pruneAnatTree(treeHuman,
              nodes = c("white matter", "sulci & spaces"),
              method = "AtNode")

#Import DSURQE labels and template
dsurqeLabels <- mincGetVolume("/projects/abeauchamp/Atlases/DSURQE/DSURQE_40micron_labels.mnc")
dsurqeAnat <- mincGetVolume("/projects/abeauchamp/Atlases/DSURQE/DSURQE_40micron_average.mnc")
dsurqeMask <- mincGetVolume("/projects/abeauchamp/Atlases/DSURQE/DSURQE_40micron_mask.mnc")
```

```{r normalization}
#Extract labels from data frames
dfLabelsMouse <- dfExprMouse %>% select(contains("Region"))
dfLabelsHuman <- dfExprHuman %>% select(contains("Region"))

#Normalize mouse data
dfExprMouse_scaled <- dfExprMouse %>% 
  select(-contains("Region")) %>% 
  as.matrix() %>% 
  scaler(axis = "rows") %>% 
  scaler(scale = FALSE, axis = "columns") %>% 
  as_tibble() %>% 
  bind_cols(dfLabelsMouse)

#Normalize human data
dfExprHuman_scaled <- dfExprHuman %>% 
  select(-contains("Region")) %>% 
  as.matrix() %>% 
  scaler(axis = "rows") %>% 
  scaler(scale = FALSE, axis = "columns") %>% 
  as_tibble() %>% 
  bind_cols(dfLabelsHuman)

#Extract genes list from mouse data (same as human)
genes <- colnames(dfExprMouse_scaled)[!str_detect(colnames(dfExprMouse_scaled), "Region")]

rm(dfExprMouse, dfExprHuman)
```

```{r initial-space-similarity}
#Compute average gene expression for 67 mouse brain regions
matExprMouse_scaled <- dfExprMouse_scaled %>% 
  select(Region = Region67,
         all_of(genes)) %>% 
  group_by(Region) %>% 
  summarise_all(mean) %>% 
  ungroup() %>% 
  column_to_rownames("Region") %>% 
  as.matrix() %>% 
  t()

#Compute average gene expression for 88 human brain regions
matExprHuman_scaled <- dfExprHuman_scaled %>% 
  select(Region = Region88,
         all_of(genes)) %>% 
  group_by(Region) %>% 
  summarise_all(mean) %>% 
  ungroup() %>% 
  column_to_rownames("Region") %>% 
  as.matrix() %>% 
  t()

#Compute human-mouse similarity matrix in gene space
matSim_H88M67_AllGenes <- buildSimilarityMatrix(x1 = matExprHuman_scaled, 
                                                x2 = matExprMouse_scaled)

#Order similarity matrix rows/columns according to brain organization
matSim_H88M67_AllGenes <- matSim_H88M67_AllGenes[,match(listLabelsMouseReordered$Region67_reordered, colnames(matSim_H88M67_AllGenes))]
matSim_H88M67_AllGenes <- matSim_H88M67_AllGenes[match(listLabelsHumanReordered$Region88_reordered, rownames(matSim_H88M67_AllGenes)),]
```

```{r import-latent-space}
#Path to MLP files
pathMLP <- str_c(pathHome, "Data/MultilayerPerceptronOutcomes/")

#Files containing latent space representations
filesMLP_Mouse <- list.files(pathMLP, pattern = "MouseTx", full.names = T)
filesMLP_Human <- list.files(pathMLP, pattern = "HumanTx", full.names = T)

#Import the mouse latent space data
listMouseLatentSpace <- map(filesMLP_Mouse,
                            function(x){
                              suppressMessages(read_csv(x)) %>% 
                                column_to_rownames("Region") %>% 
                                as.matrix() %>% 
                                t()
                            })

#Import the human latent space data
listHumanLatentSpace <- map(filesMLP_Human,
                            function(x){
                              suppressMessages(read_csv(x)) %>% 
                                column_to_rownames("Region") %>% 
                                as.matrix() %>% 
                                t()
                            })
```

```{r latent-space-similarity}
#Compute the latent space similarity matrices
listSimLatentSpace <- map2(listHumanLatentSpace, listMouseLatentSpace,
                           function(x1, x2){
                             m <- buildSimilarityMatrix(x1 = x1, x2 = x2)
                             m <- m[,match(listLabelsMouseReordered$Region67_reordered, colnames(m))]
                             m <- m[match(listLabelsHumanReordered$Region88_reordered, rownames(m)),]
                           })
names(listSimLatentSpace) <- str_c("MLP_v", 1:length(listSimLatentSpace))
rm(listMouseLatentSpace, listHumanLatentSpace)
```

```{r similarity-profiles}
#' Compute scaled similarity profiles for every column in a similarity matrix
#'
#' @param similarity A similarity matrix
#' @param scale A logical value indicating whether to scale the similarity profiles
#'
#' @return A tibble containing the scaled similarity profiles
computeSimilarityProfiles <- function(similarity, scale = FALSE){
  
  profiles <- tibble()
  for(seed in colnames(similarity)){
    temp <- similarity[, seed] %>% 
      enframe(name = "Target",
              value = "Similarity") 
      
      if(scale){
        temp <- temp %>% 
          mutate(Similarity = (Similarity - min(Similarity))/(max(Similarity) - min(Similarity)))
      }
    
    temp <- temp %>% 
      mutate(TargetRank = as.numeric(factor(Target, levels = Target[order(Similarity, decreasing = TRUE)])),
             Seed = seed)
    profiles <- bind_rows(profiles, temp)
  }
  return(profiles)
}

#Include the original homologous gene similarity matrix
listSimMats <- c(list("AllGenes" = matSim_H88M67_AllGenes),
                 listSimLatentSpace)
rm(listSimLatentSpace)

#Create data frame with similarity profiles
dfSimilarityProfiles <- map_df(.x = listSimMats,
                               .f = computeSimilarityProfiles,
                               .id = "Data",
                               scale = TRUE)
```

```{r rank-differences}
#Compute the rank at s = 0.75 for all data representations
dfRankThresholds <- dfSimilarityProfiles %>% 
  filter(Similarity >= 0.75) %>% 
  group_by(Data, Seed) %>% 
  filter(Similarity == min(Similarity)) %>% 
  ungroup() %>% 
  select(Data, Seed, TargetRank) %>% 
  spread(key = "Data", value = "TargetRank")

#Compute the differences between latent space ranks and homologous gene rank
dfRankDifferences <- dfRankThresholds %>% 
  select(contains("MLP")) %>% 
  map_dfc(function(x){x - dfRankThresholds$AllGenes}) %>% 
  mutate(Seed = dfRankThresholds$Seed) %>% 
  gather(key = "Data", value = "RankDiff", -Seed) %>% 
  mutate(Seed = fct_rev(factor(Seed, levels = listLabelsMouseReordered$Region67_reordered)))
```

# Figure 3

## Panel A: Rank-ordered similarity profile for mouse primary motor area

```{r results2-fig3-panelA-seedsim}
#Extract the chosen mouse seed
seed <- "Primary motor area"

#Some parameters
MLPsample <- "MLP_v224"
colPalette <- c("#B27700", "#00325C")
ybreaks <- seq(0, 1, 0.25)

#Data frame with MLP sample selection
dfSeedSim_Comparison <- dfSimilarityProfiles %>% 
  filter(Seed == seed,
         Data %in% c("AllGenes", MLPsample)) %>% 
  mutate(Seed = factor(Seed, levels = listLabelsMouseReordered$Region67_reordered))

#Data frame with all MLP samples
dfSeedSim_Background <- dfSimilarityProfiles %>% 
  filter(Seed == seed,
         Data != "AllGenes") %>% 
    mutate(Seed = factor(Seed, levels = listLabelsMouseReordered$Region67_reordered))

#Seed ranks at 0.75 for gene space and MLP sample
seedRanks <- c(dfRankThresholds[dfRankThresholds$Seed == seed,][["AllGenes"]],
               dfRankThresholds[dfRankThresholds$Seed == seed,][[MLPsample]])

#Rank ordered similarity plot
fig3_seedsim <- ggplot(dfSeedSim_Comparison,
                       aes(x = TargetRank,
                           y = Similarity,
                           group = Data)) +
  geom_line(data = dfSeedSim_Background,
            alpha = 0.05,
            col = "#86A6BF",
            size = 0.5) + 
  annotate(geom = "rect",
           xmin = min(seedRanks),
           xmax = max(seedRanks),
           ymin = -Inf, 
           ymax = 0.75,
           fill = "grey80",
           alpha = 0.3) + 
  annotate(geom = "segment",
           x = seedRanks,
           xend = seedRanks,
           y = -Inf,
           yend = 0.75,
           col = colPalette,
           linetype = "dashed") +
  geom_line(aes(col = Data),
            size = 1) + 
  annotate(geom = "segment",
           x = max(seedRanks),
           xend = -Inf,
           y = 0.75, 
           yend = 0.75,
           linetype = "dashed") + 
  annotate(geom = "point",
           x = seedRanks,
           y = 0.75,
           col = colPalette,
           size = 3) +
  scale_x_continuous(breaks = c(1, seq(20, 100, by = 20)),
                     limits = c(0, 89),
                     expand = c(0,0)) + 
  scale_y_continuous(limits = c(-0.05, 1.05), breaks = ybreaks) + 
  scale_colour_manual(name = "Data space",
                      labels = c("Initial gene expression space", "Latent gene expression space"),
                      values = colPalette) + 
  labs(x = "Human regions (Rank ordered)",
       y = "Scaled similarity",
       title = "Similarity profiles of the mouse primary motor area") +
  theme_bw()  +
  theme(legend.position = "bottom",
        legend.box = "horizontal",
        axis.text.x = element_text(size = 10),
        legend.title = element_blank(),
        plot.margin = margin(r = 0, b = 0, l = 0, t = 0, unit = 'inch'))

#Convert the rank ordered ggplot to a grob
fig3_seedsim_grob <- ggplotGrob(fig3_seedsim) %>% grid.force()

#Extract relevant grobs for panel A
fig3_seedsim_panel_grob <- getGrob(fig3_seedsim_grob, "panel.7-5-7-5")
fig3_seedsim_ytext_grob <- getGrob(fig3_seedsim_grob, "axis-l.7-4-7-4")
fig3_seedsim_ytitle_grob <- getGrob(fig3_seedsim_grob, "ylab-l.7-3-7-3")
fig3_seedsim_xtext_grob <- getGrob(fig3_seedsim_grob, "axis-b.8-5-8-5")
fig3_seedsim_xtitle_grob <- getGrob(fig3_seedsim_grob, "xlab-b.9-5-9-5")
fig3_seedsim_legend_grob <- getGrob(fig3_seedsim_grob, "guides.3-3-3-3")

#Plot title for panel A
fig3_seedsim_title_grob <- textGrob("Rank-ordered similarity profiles of the mouse M1",
                                    x = 0, y = 0.5,
                                    just = c("left", "centre"))
```

## Panel B: Distributions of rank differences

```{r results2-fig3-panelB-distributions}
#Prune mouse tree to 11 regions for cluster annotations
treeMouse_11 <- Clone(treeMouse)
pruneAnatTree(treeMouse_11,
              nodes = listLabelsMouseReordered$Region11,
              method = "BelowNode")

#Prune human tree to 16 regions for cluster annotations
treeHuman_16 <- Clone(treeHuman) 
pruneAnatTree(treeHuman_16,
              nodes = listLabelsHumanReordered$Region16,
              method = "BelowNode")

#Get colour annotation from trees
coloursMouse <-  treeMouse_11$Get("color_hex_triplet", filterFun = isLeaf)
indOrderMouseColours <- match(listLabelsMouseReordered$Region11_reordered, names(coloursMouse))
coloursMouse <- coloursMouse[indOrderMouseColours]

coloursHuman <- treeHuman_16$Get("color_hex_triplet", filterFun = isLeaf)
indOrderHumanColours <- match(listLabelsHumanReordered$Region16_reordered, names(coloursHuman))
coloursHuman <- coloursHuman[indOrderHumanColours]

annotation_colours <- list(MouseRegion = coloursMouse,
                           HumanRegion = coloursHuman)

#Create a set of modified colours for clearer visualization. 
#To be used for text colouring mostly
annotation_colours_mod <- annotation_colours
annotation_colours_mod$MouseRegion["Isocortex"] <- "#68EC68"
annotation_colours_mod$MouseRegion["Cerebellar cortex"] <- "#E6E67B"
annotation_colours_mod$MouseRegion["Cerebellar nuclei"] <- "#E6E67B"

annotation_colours_mod$HumanRegion["insula"] <- "#E6E658"
annotation_colours_mod$HumanRegion["diencephalon"] <- "#85DD63"
annotation_colours_mod$HumanRegion["pons"] <- "#00E0A4"

#Create data frame with summary stats for MLP samples
dfRankDifferencesSummary <- dfRankDifferences %>% 
  mutate(NonPositive = ifelse(RankDiff <= 0, TRUE, FALSE)) %>% 
  group_by(Seed) %>% 
  summarise(RankDiffMean = mean(RankDiff),
            RankDiffSd = sd(RankDiff),
            RankDiffLower = RankDiffMean - 2*RankDiffSd,
            RankDiffUpper = RankDiffMean + 2*RankDiffSd,
            PropNonPos = sum(NonPositive)/n()) %>% 
  ungroup() 

#Create a data frame containing mouse colours for y-axis annotations
dfMouseColours <- annotation_colours_mod$MouseRegion %>% 
  enframe(name = "Region11",
          value = "Colour") %>% 
  inner_join(dfLabelsMouse %>% 
               select(Region11, Region67) %>% 
               distinct(),
             by = "Region11") %>% 
  mutate(Region67 = fct_rev(factor(Region67, levels = listLabelsMouseReordered$Region67_reordered)),
         Region11 = factor(Region11, levels = listLabelsMouseReordered$Region11_reordered))

#Join summary and colours data frames
dfRankDifferencesSummary <- inner_join(dfRankDifferencesSummary,
                                       dfMouseColours,
                                       by = c("Seed" = "Region67"))

#Parameters
ybreaks <- seq(-50, 50, by = 25)

#Rank differences distribution plot
fig3_distributions <- ggplot(dfRankDifferencesSummary, aes(x = Seed, y = RankDiffMean, col = Region11)) + 
  geom_hline(yintercept = 0,
             linetype = "dashed") + 
  geom_pointrange(aes(ymin = RankDiffLower,
                      ymax = RankDiffUpper),
                  size = 0.6,
                  fatten = 2) + 
  coord_flip() + 
  scale_x_discrete(position = "top") + 
  scale_y_continuous(breaks = ybreaks) + 
  scale_colour_manual(values = annotation_colours_mod$MouseRegion) + 
  labs(x = "Mouse region",
       y = "Rank difference",
       title = "Difference in rank at s = 0.75 between gene space and latent spaces") + 
  theme_bw() + 
  theme(axis.text.y = element_text(size = 6,
                                   colour = dfRankDifferencesSummary$Colour,
                                   face = "bold"),
        legend.position = "none",
        plot.margin = margin(r = 0, b = 0, l = 0, t = 0, unit = 'inch'))
fig3_distributions_grob <- ggplotGrob(fig3_distributions) %>% grid.force()

#Extract relevant grobs for panel B
fig3_distributions_panel_grob <- getGrob(fig3_distributions_grob, "panel.7-5-7-5")
fig3_distributions_ytext_grob <- getGrob(fig3_distributions_grob, "axis-r.7-6-7-6")
fig3_distributions_ytitle_grob <- getGrob(fig3_distributions_grob, "ylab-r.7-7-7-7")
fig3_distributions_xtext_grob <- getGrob(fig3_distributions_grob, "axis-b.8-5-8-5")
fig3_distributions_xtitle_grob <- getGrob(fig3_distributions_grob, "xlab-b.9-5-9-5")

#Plot title for panel B
fig3_distributions_title_grob <- textGrob("Difference in rank at a scaled similarity of 0.75",
                                          x = 0, y = 0.5,
                                          just = c("left", "centre"),
                                          gp = gpar(lineheight = 0.8))
```

## Panel C: Proportions slice series

```{r results2-fig3-panelC-proportions}
#Generate a mouse tree with 67 leaf nodes
treeMouse_67 <- Clone(treeMouse)
pruneAnatTree(treeMouse_67,
              nodes = listLabelsMouseReordered$Region67_reordered,
              method = "BelowNode")

#Create a mouse atlas with 67 regions
atlasMouse_67 <- hanatToAtlas(treeMouse_67, mincArray(dsurqeLabels))
dfAtlasMouse_67 <- hanatToAtlasDefs(treeMouse_67)

#Create MINC array for proportions non-positive
fig3_ss_array <- mincBuildArray(values = dfRankDifferencesSummary$PropNonPos,
                                values.names = dfRankDifferencesSummary$Seed,
                                labels = atlasMouse_67,
                                defs = dfAtlasMouse_67)

#Slice series palette
fig3_ss_palette <- viridis(n = 255)

#Generate base slice series
fig3_ss_grob <- sliceSeries(nrow = 3, ncol = 4, begin = 70, end = 330) %>% 
  anatomy(mincArray(dsurqeAnat), low = 700, high = 1400) %>% 
  overlay(mincArray(fig3_ss_array), low = 0, high = 1, col = fig3_ss_palette) %>% 
  grobify()

#Create custom slice series legend
#Legend scale
fig3_ss_legend_palette <- fig3_ss_palette
fig3_ss_legend_palette <- fig3_ss_legend_palette[length(fig3_ss_legend_palette):1]
fig3_ss_legend_scale_width <- 0.3
fig3_ss_legend_scale_height <- 1
fig3_ss_legend_scale_x <- 0.95
fig3_ss_legend_scale_y <- 0.5
fig3_ss_legend_scale_grob <- rasterGrob(fig3_ss_legend_palette,
                                        x = fig3_ss_legend_scale_x,
                                        y = fig3_ss_legend_scale_y,
                                        width = fig3_ss_legend_scale_width,
                                        height = fig3_ss_legend_scale_height,
                                        just = c("right", "centre"))

#Legend text
fig3_ss_legend_text <- c(min(fig3_ss_array), max(fig3_ss_array))
fig3_ss_legend_text_x <- fig3_ss_legend_scale_x - fig3_ss_legend_scale_width - 0.1
fig3_ss_legend_text_y <- fig3_ss_legend_scale_y + 0.5*c(-1*fig3_ss_legend_scale_height,
                                                        fig3_ss_legend_scale_height)
fig3_ss_legend_text_grob <- textGrob(fig3_ss_legend_text,
                                     x = fig3_ss_legend_text_x,
                                     y = fig3_ss_legend_text_y)

#Combine slice series legend grobs
fig3_ss_legend_grob <- gTree(children = gList(fig3_ss_legend_scale_grob,
                                              fig3_ss_legend_text_grob))

#Plot title for panel C
fig3_ss_title_grob <- textGrob("Proportion of latent spaces featuring improved rank",
                               x = 0, y = 0.5,
                               just = "left")
```

## Figure 3 grob

```{r results2-fig3-grob, fig.width = 12, fig.height = 10}
#Generate the main figure grob
fig3_plots_grob <- arrangeGrob(textGrob("A.", gp = gpar(fontsize = 14, fontface = "bold")), #1
                               fig3_seedsim_title_grob, #2
                               textGrob("B.", gp = gpar(fontsize = 14, fontface = "bold")), #3
                               fig3_distributions_title_grob, #4
                               fig3_seedsim_ytitle_grob, #5
                               fig3_seedsim_ytext_grob, #6
                               fig3_seedsim_panel_grob, #7
                               fig3_distributions_panel_grob,#8
                               fig3_distributions_ytext_grob, #9
                               fig3_distributions_ytitle_grob, #10
                               fig3_seedsim_xtext_grob, #11
                               fig3_seedsim_xtitle_grob, #12
                               fig3_seedsim_legend_grob, #13
                               textGrob("C.", gp = gpar(fontsize = 14, fontface = "bold")), #14
                               fig3_ss_title_grob, #15
                               fig3_ss_legend_grob, #16
                               fig3_ss_grob, #17
                               fig3_distributions_xtext_grob, #18
                               fig3_distributions_xtitle_grob, #19
                               layout_matrix = rbind(c(NA, NA, NA, NA, NA, NA, NA, NA, NA),
                                                     c(NA,  1, NA,  2,  3,  4,  4,  4, NA),
                                                     c(NA,  5,  6,  7, NA,  8,  9, 10, NA),
                                                     c(NA, NA, NA, 11, NA,  8,  9, 10, NA),
                                                     c(NA, NA, NA, 12, NA,  8,  9, 10, NA),
                                                     c(NA, NA, NA, 13, NA,  8,  9, 10, NA),
                                                     c(NA, NA, NA, NA, NA,  8,  9, 10, NA),
                                                     c(NA, 14, NA, 15, NA,  8,  9, 10, NA),
                                                     c(NA, 16, 16, 17, NA,  8,  9, 10, NA),
                                                     c(NA, NA, NA, NA, NA, 18, NA, NA, NA),
                                                     c(NA, NA, NA, NA, NA, 19, NA, NA, NA),
                                                     c(NA, NA, NA, NA, NA, NA, NA, NA, NA)),
                               widths = unit(c(0.1, 0.3, 0.4, 4.5, 0.45, 2.2, 1.5, 0.4, 0.1), 'inch'),
                               heights = unit(c(0.25, 0.3, 2, 0.25, 0.25, 0.4, 0.1, 0.4, 2.7, 0.2, 0.3, 0.25), 'inch'))

#Empty rectangle grob to form the border
fig3_border_grob <- rectGrob(gp = gpar(fill = NA))

#Import AMBA legend slice
load(str_c(pathHome, "Draft/", "Version5/", "AMBALegendSlice.RData"))

#Width and height for the atlas legend slice
sliceAMBA_width <- 0.9
sliceAMBA_height <- sliceAMBA_width/sliceAMBA_asp

#New viewport for the atlas legend slice
sliceAMBALegend_grob$vp <- viewport(x = unit(9.5, 'inch'),
                                    y = unit(1.4, 'inch'),
                                    width = unit(sliceAMBA_width, 'inch'),
                                    height = unit(sliceAMBA_height,'inch'))

#Figure 3 grob
fig3_grob <- gTree(children = gList(fig3_plots_grob,
                                    fig3_border_grob,
                                    sliceAMBALegend_grob))

fig3_width <- 10
fig3_height <- 7.4

grid.newpage()
grid.draw(fig3_grob)
```

```{r results2-fig3-write}
fileout <- str_c(pathHome, "Draft/", "Version5/", "Figure3.RData")
save(fig3_grob,
     fig3_width,
     fig3_height,
     file = fileout)
```


# In-text quantification

```{r results2-fig3-intext-1}
#Number of latent spaces
nMLP <- length(unique(dfRankDifferences$Data))

#Data frame containing proportions non-positive rank differences
temp <- dfRankDifferences %>% 
  mutate(NonPositive = ifelse(RankDiff <= 0, TRUE, FALSE),
         Positive = !NonPositive) %>%  
  group_by(Seed) %>% 
  summarise(PropPositive = sum(Positive)/nMLP,
            PropNonPositive = sum(NonPositive)/nMLP,
            RankDiffMean = mean(RankDiff),
            RankDiffSd = sd(RankDiff)) %>% 
  ungroup() 

#Number and proportion of regions having mean rank difference less than or equal to zero
temp %>% 
  summarise(nregions = sum(RankDiffMean <= 0),
            fracregions = nregions/n())
```

```{r results2-fig3-intext-2}
#Proportion of regions having non-positive rank difference in greater than 80% of latent spaces
temp %>% 
  summarise(sum(PropNonPositive >= 0.80)/n())
```


```{r results2-fig3-intext-3}
#Summary statistics for specific mouse regions
temp %>% 
  filter(Seed %in% c("Main olfactory bulb", "Accessory olfactory bulb", "Cerebellar nuclei"))
```

```{r results2-fig3-intext-4}
#Mouse regions in the isocortex
structsCx <- dfLabelsMouse %>%
  select(Region11, Region67) %>% 
  distinct() %>% 
  filter(Region11 == "Isocortex") %>% 
  pull(Region67)

#Number and proportion of isocortical regions with non-positive rank differences in at least 96% of latent spaces
temp %>% 
  filter(Seed %in% structsCx) %>% 
  summarise(nregions = sum(PropNonPositive >= 0.96),
            fracregions = nregions/n())
```

```{r results2-fig3-intext-5}
#Mouse regions in the cerebellar cortex
structsCb <- dfLabelsMouse %>%
  select(Region11, Region67) %>% 
  distinct() %>% 
  filter(Region11 == "Cerebellar cortex") %>% 
  pull(Region67)

#Number and proportion of cerebellar cortical regions with non-positive rank differences in at least 96% of latent spaces
temp %>% 
  filter(Seed %in% structsCb) %>% 
  summarise(nregions = sum(PropNonPositive >= 0.96),
            fracregions = nregions/n())
```

```{r results2-fig3-intext-6}
#Summary statistics for thalamus
temp %>% 
  filter(Seed == "Thalamus")
```

```{r results2-fig3-intext-7}
#Mouse regions in the sub-cortex
structsSubCx <- dfLabelsMouse %>% 
  select(Region11, Region67) %>% 
  distinct() %>% 
  filter(Region11 %in% c("Cerebral nuclei", "Diencephalon", "Midbrain", "Pons", "Medulla")) %>% 
  pull(Region67)

#Mean and sd of rank differences for subcortical regions
dfRankDifferences %>% 
  filter(Seed %in% structsSubCx) %>% 
  summarise(mean(RankDiff),
            sd(RankDiff))
```
